Allow config authors to always use Prefix

lsm5 opened the pull request:
Previously an unset Location was only allowed for wildcarded Prefixes.
This commit will allow any valid Prefix.

Fixes: https://github.com/containers/image/pull/1191#discussion_r610622495

Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

@vrothberg @mtrmac @ashcrow  PTAL

vrothberg reviewed (commented):
Needs tests and changes to docs.  AFAIR the docs state the location must be set, and we need to explain how rewriting works when location is empty.

The tests should make sure that rewriting the reference will return the input.

lsm5 commented:
Done. PTAL.

lsm5 commented:
@mtrmac removing the entire check for empty location from rewriteReference seemed to cause failures in `TestRewriteReferenceFailedDuringParseNamed` . So, let me know if the FIXME comment still needs to exist there.

vrothberg commented on docs/containers-registries.conf.5.md:
```diff
@@ -82,7 +82,7 @@ location = "internal-registry-for-example.net/bar"
 
-With a `prefix` containing a wildcard in the format: "*.example.com" for subdomain matching,
+With any valid `prefix` with or without a wildcard in the format: "*.example.com" for subdomain matching,
 the location can be empty. In such a case,
```
This sentence seems a bit too complicated. I suggest: "The `location` can be empty if the `prefix` is set."

vrothberg commented on pkg/sysregistriesv2/system_registries_v2.go:
```diff
@@ -357,21 +350,26 @@ func (config *V2RegistriesConf) postProcessRegistries() error {
-			}
-			reg.Prefix = reg.Location
-		} else {
+		// Allow config authors to always use Prefix.
```
I don't find the comment helpful. I'd prefer a comment stating that either prefix OR location OR both must be set.

vrothberg reviewed (commented):
I'd love a test with at least one registries.conf file with an empty location and set prefix.

rhatdan commented:
@lsm5 any update on this PR?

lsm5 commented:
@rhatdan working on it..

lsm5 commented:
@mtrmac @vrothberg PTAL

vrothberg reviewed (commented):
LGTM
@mtrmac PTAL

lsm5 commented:
@mtrmac ping PTAL

vrothberg commented:
> @mtrmac ping PTAL

On PTO until next week.

lsm5 committed:
Allow config authors to always use Prefix

Previously an unset Location was only allowed for wildcarded Prefixes.
This commit will allow configs with emtpy location or empty prefix, so
long as the other is valid.

Remove wildcard prefix check for empty location in rewriteReference.

Fixes:
1. https://github.com/containers/image/pull/1191#discussion_r610622495
2. https://github.com/containers/image/pull/1191#discussion_r610621608

Co-authored-by: Miloslav Trmač <mitr@redhat.com>
Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

lsm5 commented:
@mtrmac PTAL

mtrmac commented on pkg/sysregistriesv2/system_registries_v2.go:
```diff
@@ -357,21 +350,27 @@ func (config *V2RegistriesConf) postProcessRegistries() error {
-		} else {
+		// Prefix and Location cannot both be empty.
+		// Either one at least must be set.
+		if reg.Prefix != "" {
```
This could be simpler as
```go
if reg.Prefix == "" && reg.Location == "" { fail }
if reg.Location != "" { reg.Location, err = parseLocation(…); if err … }
if reg.Prefix == "" { reg.Prefix = reg.Location }
else { reg.Prefix, err = parseLocation(…); if err … } 
```

or perhaps
```go
// Validate individual fields
if reg.Location != "" { reg.Location, err = parseLocation(…); if err … }
if reg.Prefix != "" { reg.Prefix, err = parseLocation(…); if err … } 
// Validate the combination, normalize to have reg.Prefix always set
if reg.Prefix == "" {
    if reg.Location == "" { fail }
    else { reg.Prefix = reg.Location }
}
```

Either way, please remove the `parseLocation` call above this line, and make `parseLocation` fail on `""` again.

mtrmac commented on pkg/sysregistriesv2/system_registries_v2_test.go:
```diff
@@ -383,6 +383,12 @@ func TestFindRegistry(t *testing.T) {
 	assert.Equal(t, "empty-prefix.com", reg.Location)
 
+	reg, err = FindRegistry(sys, "empty-location.set-prefix.example.com/namespace/repo")
+	assert.Nil(t, err)
+	assert.NotNil(t, reg)
```
(Non-blocking: These should be `require.`. OTOH this is consistent with the rest, and only matters on test failures.)

mtrmac replied:
I honestly don’t think this is worth much worrying about; if you find updating pre-existing test cases fun, sure, go for it, but it’s not work that results in anything users could notice and consider valuable while _using_ the software. As far as “clean code” goes, `require` is a bit more correct but consistency is also nice, so to me that’s not a strong reason to prefer either.

(Just to be explicit: `require` should be used for testing prerequisites that must be true for other checks to be even valid to do, e.g.
```go
ptr, err := testSubject()
require.NoError(t, err)
assert.Equal(t, …, *ptr) // would crash if testSubject returned (nil, err) and the above were assert.NoError(…)
```

But there is a _minimal_ practical difference; with `assert.NoError(t, err)` the test crashes with a backtrace, with `require.NoError(t, err)` the test fails with an error message containing `err`. It doesn’t matter at all if the test completely succeeds. If the test fails because `err` is set, in both cases the test fails, it’s easy enough to identify the location of the failure, and probably something about `testSubject` needs to be updated. `require` is only better in that it includes the actual error text in the test failure output, and that could help with a inconsistent rarely-occurring failure in CI.)

mtrmac reviewed (commented):
Thanks for the PR, and I apologize for the late review.

ACK overall; a few more simplifications seem possible.

jankaluza commented:
Hi, and thank you for your contribution!

We’ve recently migrated this repository into a new **monorepo**: [containers/container-libs](https://github.com/containers/container-libs) along with other repositories

As part of this migration, this repository is no **longer accepting new Pull-Requests** and therefore this Pull-Request is being closed.

Thank you very much for your contribution. We would appreciate your continued help in migrating this PR to the new [container-libs repository](https://github.com/containers/container-libs). Please let us know if you are facing any issues.

You can read more about the migration and the reasoning behind it in our blog post: [Upcoming migration of three containers repositories to monorepo](https://blog.podman.io/2025/08/upcoming-migration-of-three-containers-repositories-to-monorepo/).

Thanks again for your work and for supporting the containers ecosystem!
